library/src/tests/diagnostics.rs

```diff
@@ -156,3 +156,117 @@ fn check_stop_counting_function_without_start() {
+
+            StartCountingQubits();
+            use q = Qubit();
+            StopCountingQubits()
+        }",
+        &Value::Int(1),
+    );
+}
+
+#[test]
+fn check_start_counting_qubits_for_tuple_allocation() {
+    test_expression(
+        "{
+            import Microsoft.Quantum.Diagnostics.StartCountingQubits;
+            import Microsoft.Quantum.Diagnostics.StopCountingQubits;
+
+            StartCountingQubits();
+            use (q0, q1) = (Qubit(), Qubit());
+            StopCountingQubits()
+        }",
+        &Value::Int(2),
+    );
+}
+
+#[test]
+fn check_start_counting_qubits_for_array_allocation() {
+    test_expression(
+        "{
+            import Microsoft.Quantum.Diagnostics.StartCountingQubits;
+            import Microsoft.Quantum.Diagnostics.StopCountingQubits;
+
+            StartCountingQubits();
+            use qs = Qubit[2];
+            StopCountingQubits()
+        }",
+        &Value::Int(2),
+    );
+}
+
+#[test]
+fn check_start_counting_qubits_after_allocation_gives_zero() {
+    test_expression(
+        "{
+            import Microsoft.Quantum.Diagnostics.StartCountingQubits;
+            import Microsoft.Quantum.Diagnostics.StopCountingQubits;
+
+            use q = Qubit();
+            StartCountingQubits();
+            StopCountingQubits()
+        }",
+        &Value::Int(0),
+    );
+}
+
+#[test]
+fn check_start_counting_qubits_sees_same_qubit_as_single_count() {
+    test_expression(
+        "{
+            import Microsoft.Quantum.Diagnostics.StartCountingQubits;
+            import Microsoft.Quantum.Diagnostics.StopCountingQubits;
+
+            StartCountingQubits();
+            {
+                use q = Qubit();
+            }
+            {
+                use q = Qubit();
+            }
+            StopCountingQubits()
+        }",
+        &Value::Int(1),
+    );
+}
+
+#[test]
+fn check_start_counting_qubits_works_with_manual_out_of_order_allocation_release() {
+    test_expression(
+        "{
+            import Microsoft.Quantum.Diagnostics.StartCountingQubits;
+            import Microsoft.Quantum.Diagnostics.StopCountingQubits;
+            import QIR.Runtime.__quantum__rt__qubit_allocate;
+            import QIR.Runtime.__quantum__rt__qubit_release;
+
+            let (q0, q1, q2) = (__quantum__rt__qubit_allocate(), __quantum__rt__qubit_allocate(), __quantum__rt__qubit_allocate());
+            StartCountingQubits();
+            __quantum__rt__qubit_release(q2);
+            use q = Qubit();
+            __quantum__rt__qubit_release(q0);
+            __quantum__rt__qubit_release(q1);
+            use qs = Qubit[2];
+            StopCountingQubits()
+        }",
+        &Value::Int(3),
+    );
+}
+
+#[test]
+fn check_stop_counting_qubits_without_start_returns_negative_one() {
+    test_expression(
+        "{
+            import Microsoft.Quantum.Diagnostics.StopCountingQubits;
+
+            StopCountingQubits()
+        }",
+        &Value::Int(-1),
```

Comment:
This looks odd... I feel that if we don't count something, the natural count of that is 0, why a negative value?

Reply:
Per conversation on the other PR, this is now an error instead.